File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
function lengthOfLongestSubstring(s) {
    let longest = 0;
    let counter = new Map();
    let l = 0;
    for (var right = 0; right < s.length; ++right) {
        const cur = s[right];
        counter.set(cur, (counter.has(cur))? counter.get(cur)+1: 1);
        while (counter.get(cur) > 1) {
            counter.set(s[l], counter.get(s[l])-1);
            ++l;
        }
        longest = Math.max(longest, right-l+1);
    }
    return longest;
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = set()
        longest = 0
        left, right = 0, 0
        while right < len(s):
            if s[right] not in chars:
                chars.add(s[right])
            else:
                while s[left] != s[right]:
                    chars.remove(s[left])
                    left += 1
                left += 1
            longest = max(longest, right - left + 1)
            right += 1

        return longest
        
        # keep track of the window's chars in a set
        # keep track of the longest substring
        # use sliding window
        # if the char is not in the set, add it
        # if the char is in the set, increment left until you hit the same char
        # update the set as needed